src/persist-client/src/impl/state.rs

```diff
@@ -568,4 +570,37 @@ mod tests {
             .collect::<Vec<_>>();
         assert_eq!(actual, vec!["key1", "key2", "key3"]);
     }
+
+    #[test]
+    fn compare_and_append() {
+        mz_ore::test::init_logging();
+        let mut state = State::<String, String, u64, i64>::new(ShardId::new()).collections;
+
+        // State is initally empty.
+        assert_eq!(state.trace.len(), 0);
+
+        // Cannot insert a batch with a lower != current shard upper.
+        assert_eq!(
+            state.compare_and_append(&["key1".to_owned()], &desc(1, 2)),
+            Break(Ok(Upper(Antichain::from_elem(0))))
+        );
+
+        // Insert a batch to move the shard upper to a nonzero value.
+        assert_eq!(state.compare_and_append(&[], &desc(0, 5)), Continue(()));
+
+        // Cannot insert a batch with a upper less than the lower.
+        assert_eq!(
+            state.compare_and_append(&["key1".to_owned()], &desc(5, 4)),
+            Break(Err(InvalidBounds {
+                lower: Antichain::from_elem(5),
+                upper: Antichain::from_elem(4)
+            }))
+        );
+
+        // Can insert a batch with an upper equal to lower.
```

Comment:
that's a bug!

Reply:
cool glad we're on the same page. I feel like theres at least two ways to resolve this and I don't have a preference at all between them.

1. Ban batches over empty time intervals for `compare_and_append`. Forces the user to commit to strict progress of time and is simpler to reason about.

2. Only allow empty batches (zero rows of data) over empty time intervals. This doesn't let the user communicate progress (because the shard upper is not advancing) but it does allow for some form of periodic heartbeat thats not coupled to the rate at which the times are advancing. It adds an extra branch through the code however.

My weak preference is to impl 1 and then revisit if we end up needing 2. wdyt?

Reply:
Glad I asked! I'll do 2

Reply:
done.